src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -0,0 +1,581 @@
+    internal partial struct TarHeader
+    {
+        private const string UstarPrefixFormat = "{0}/{1}"; // "prefix/name"
+
+        // Attempts to read all the fields of the next header.
+        // Throws if end of stream is reached or if any data type conversion fails.
+        // Returns true if all the attributes were read successfully, false otherwise.
+        internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
+        {
+            // The four supported formats have a header that fits in the default record size
+            byte[] rented = ArrayPool<byte>.Shared.Rent(minimumLength: TarHelpers.RecordSize);
+
+            Span<byte> buffer = rented.AsSpan(0, TarHelpers.RecordSize); // minimumLength means the array could've been larger
+            buffer.Clear(); // Rented arrays aren't clean
+
+            TarHelpers.ReadOrThrow(archiveStream, buffer);
+
+            try
+            {
+                // Confirms if v7 or pax, or tentatively selects ustar
+                if (!TryReadCommonAttributes(buffer))
+                {
+                    return false;
+                }
+
+                // Confirms if gnu, or tentatively selects ustar
+                ReadMagicAttribute(buffer);
+
+                if (_format != TarFormat.V7)
+                {
+                    // Confirms if gnu
+                    ReadVersionAttribute(buffer);
+
+                    // Fields that ustar, pax and gnu share identically
+                    ReadPosixAndGnuSharedAttributes(buffer);
+
+                    Debug.Assert(_format is TarFormat.Ustar or TarFormat.Pax or TarFormat.Gnu);
+                    if (_format == TarFormat.Ustar)
+                    {
+                        ReadUstarAttributes(buffer);
+                    }
+                    else if (_format == TarFormat.Gnu)
+                    {
+                        ReadGnuAttributes(buffer);
+                    }
+                    // In PAX, there is nothing to read in this section (empty space)
+                }
+
+                ProcessDataBlock(archiveStream, copyData);
+
+                return true;
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(rented);
+            }
+        }
+
+        // Reads the elements from the passed dictionary, which comes from the first global extended attributes entry,
+        // and inserts or replaces those elements into the current header's dictionary.
+        // If any of the dictionary entries use the name of a standard attribute (not all of them), that attribute's value gets replaced with the one from the dictionary.
+        // Unlike the historic header, numeric values in extended attributes are stored using decimal, not octal.
+        // Throws if any conversion from string to the expected data type fails.
+        internal void ReplaceNormalAttributesWithGlobalExtended(IReadOnlyDictionary<string, string> gea)
```

Comment:
Where does this come from?  If this is the `ReadOnlyDictionary<>` we allocate to wrap an internal `Dictionary<>`, can this operate on that `Dictionary<>` instead?  It'll make operations below, like enumerating, much cheaper.